core-bundle/src/Image/Studio/FigureBuilder.php

```diff
@@ -0,0 +1,524 @@
+     *
+     * @var array<string, string|null>
+     */
+    private $additionalLinkAttributes = [];
+
+    /**
+     * User defined light box resource or url (overwriting the default).
+     *
+     * @var string|ImageInterface|null
+     */
+    private $lightBoxResourceOrUrl;
+
+    /**
+     * User defined light box size configuration (overwriting the default).
+     *
+     * @var mixed|null
+     */
+    private $lightBoxSizeConfiguration;
+
+    /**
+     * User defined light box group identifier (overwriting the default).
+     *
+     * @var string|null
+     */
+    private $lightBoxGroupIdentifier;
+
+    /**
+     * Determines if a light box (or "fullsize") image should be created.
+     *
+     * @var bool
+     */
+    private $enableLightBox;
+
+    /**
+     * @internal Use the `contao.image.studio` factory to get an instance of this class.
+     */
+    public function __construct(ContainerInterface $locator)
+    {
+        $this->locator = $locator;
+    }
+
+    /**
+     * Set the image resource from a FilesModel.
+     */
+    public function fromFilesModel(FilesModel $filesModel): self
+    {
+        if ('file' !== $filesModel->type) {
+            throw new \InvalidArgumentException("DBAFS item '{$filesModel->path}' is not a file.");
+        }
+
+        $this->filePath = Path::makeAbsolute($filesModel->path, $this->projectDir());
+        $this->filesModel = $filesModel;
+
+        return $this;
+    }
+
+    /**
+     * Set the image resource from a tl_files uuid.
+     */
+    public function fromUuid(string $uuid): self
+    {
+        $filesModel = $this->filesModelAdapter()->findByUuid($uuid);
+
+        if (null === $filesModel) {
+            throw new \InvalidArgumentException("DBAFS item with uuid '$uuid' could not be found.");
+        }
+
+        return $this->fromFilesModel($filesModel);
+    }
+
+    /**
+     * Set the image resource from a tl_files id.
+     */
+    public function fromId(int $id): self
+    {
+        $filesModel = $this->filesModelAdapter()->findByPk($id);
+
+        if (null === $filesModel) {
+            throw new \InvalidArgumentException("DBAFS item with id '$id' could not be found.");
+        }
+
+        return $this->fromFilesModel($filesModel);
+    }
+
+    /**
+     * Set the image resource from an absolute or relative path.
+     *
+     * @param bool $autoDetectDbafsPaths Set to false to skip searching for a FilesModel
+     */
+    public function fromPath(string $path, bool $autoDetectDbafsPaths = true): self
+    {
+        $projectDir = $this->projectDir();
+
+        // Make sure path is absolute and in a canonical form
+        $path = Path::isAbsolute($path) ?
+            Path::canonicalize($path) :
+            Path::makeAbsolute($path, $projectDir);
+
+        // Only check for a FilesModel if requested if resource is inside upload path
+        if ($autoDetectDbafsPaths && Path::isBasePath(Path::join($projectDir, $this->uploadPath()), $path)) {
```

Comment:
Could you elaborate what you mean?

Reply:
Is `Dbafs::shouldBeSynchronized()` essentially saying: this is a valid DBAFS resource and nothing about it's current sync state?

(Should this be changed as well in your opinion?  https://github.com/contao/contao/blob/b4460960d6d7dc35bf4a87bcb7a503bce1df0d6d/core-bundle/src/Image/ImageFactory.php#L305-L307)

The thing I don't like with  `Dbafs::shouldBeSynchronized()` is that it performs `is_file()` / `is_dir()` checks. And the condition you commented on is only there for performance because string pattern matching is cheap. :shrug:

Reply:
yeah, I don't know exactly. It's just that there is an existing method for exactly this case. I agree its not optimized. But if its true that we don't care about _false positives_ (running `FilesModel::findByPath` on excluded folders), maybe we shouldn't check the path at all?

Reply:
This was commented on originally by @ausi in https://github.com/contao/contao/pull/1753#discussion_r435108259. I'm fine with any solution.